packages/forms/src/directives/validators.ts

```diff
@@ -167,14 +167,14 @@ export const MAX_VALIDATOR: StaticProvider = {
   selector:
       'input[type=number][max][formControlName],input[type=number][max][formControl],input[type=number][max][ngModel]',
   providers: [MAX_VALIDATOR],
-  host: {'[attr.max]': 'max ? max : null'}
+  host: {'[attr.max]': 'max ?? 0'}
 })
 export class MaxValidator extends AbstractValidatorDirective implements OnChanges {
   /**
    * @description
    * Tracks changes to the max bound to this directive.
    */
-  @Input() max!: string|number;
+  @Input() max!: string|number|null;
```

Comment:
I think we should explicitly allow `undefined` in addition to `null`.

```suggestion
  @Input() max: string|number|null|undefined;
```

Reply:
@PowerKiKi could you please elaborate a little more on why you think `undefined` should be included?

Adding `undefined` would allow for syntax like this: `<input ngModel max>`, i.e. where `min` and `max` attributes would be present without any value and that would not trigger a template type checking error.

Reply:
@PowerKiKi as per conversation in this ticket we are only going to support only null type here, If you have use case of undefined it will be easy to include your suggested change.

Please read conversation here.
[https://github.com/angular/angular/pull/42378#discussion_r641073378](https://github.com/angular/angular/pull/42378#discussion_r641073378)

Reply:
My use-case for undefined was briefly described in #42267. I have a custom component that allows me to configure an `<input type="number">` (amongst other things). Usage for the custom component would be like:

```html
<my-cmp [config]="config"></my-cmp>
```

The component being declared with something like:

```ts
type Config = {min?: number, otherThings?: string};

@Component({
    selector: 'my-cmp',
    template: '<input type="number" [min]="config.min">',
})
class MyCmp {
    @Input() config: Config;
}
```

Because I don't want to force to always specify the entire configuration, some (or all) fields are optional, hence `undefined`-able. But then I cannot just pass them as shown above. It would not compile in strict mode. I would have to jump through a few hoops.

A possibly annoying hoop, albeit common, would be to merge the incoming config, with default values and storing the result as a complete non-undefined config. Probably something like that instead:


```diff
 type Config = {min?: number, otherThings?: string};

 @Component({
     selector: 'my-cmp',
-    template: '<input type="number" [min]="config.min">',
+    template: '<input type="number" [min]="completeConfig.min">',
 })
 class MyCmp {
+    private readonly defaultConfig: Required<Config> = {min: null, otherThings: ''};
+    public completeConfig: Required<Config> = this.defaultConfig;
-    @Input() config: Config;
+    @Input() set config(value: Config) { this.completeConfig = {...this.defaultConfig, ...value}; }
 }
```

Another, simpler, hoop would be to spread default values all over the template like so:

```diff
 type Config = {min?: number, otherThings?: string};

 @Component({
     selector: 'my-cmp',
-    template: '<input type="number" [min]="config.min">',
+    template: '<input type="number" [min]="config.min ?? null">',
 })
 class MyCmp {
     @Input() config: Config;
 }
```

I am not entirely sure which was would have my preference. First one feels more structured and readable, but a bit more boilerplate.

So I guess _there are_ solutions to my use-case even without angular core supporting `undefined`, but I wanted to be sure that this use-case was considered before making a decision.

The other, problematic, use-case of allowing  `<input [formControl]="ctrl" min>`, as described in https://github.com/angular/angular/pull/42378#discussion_r641073378 did not cross my mind until now. And I suppose it would be enough to justify me having to jump the hoops I described. WDTY ?

Reply:
@PowerKiKi thanks for the reply and describing the use-case. I think that we'd need to make sure that there is an explicit value provided in cases when you want to dynamically disable min/max validator.

Another risk of having `undefined` support might be that it might hide real errors in templates. For example if you have `<input type="number" [min]="config.min">`, the `config.min` field might be missing on the `config` by mistake, thus being `undefined` in that expression and the Forms logic would treat it as "no validator needed here" instead of raising an error.

Reply:
> it might hide real errors in templates

Excellent point ! 👍

 Let's go ahead without support of `undefined`, and I'll deal with it in my own code with one of the workarounds mentioned earlier.